src/client/array/dc_array.c

```diff
@@ -2287,30 +2299,33 @@ add_record(daos_handle_t oh, daos_handle_t th, struct set_size_props *props, d_l
 
 	params->akey_val = '0';
 	params->next = NULL;
-	params->user_sgl_used = false;
 	params->dkey_val = props->dkey_val;
 	d_iov_set(dkey, &params->dkey_val, sizeof(uint64_t));
 
-	/** set memory location */
+	/** set memory location - freed in free_set_size_cb (already registered) */
 	D_ALLOC(props->val, props->cell_size);
+	if (props->val == NULL)
+		D_GOTO(err, rc = -DER_NOMEM);
 	sgl->sg_nr = 1;
-	D_ALLOC_PTR(sgl->sg_iovs);
+	sgl->sg_iovs = &params->sg_iov;
 	d_iov_set(&sgl->sg_iovs[0], props->val, props->cell_size);
+	params->user_sgl_used = true;
```

Comment:
same, the user_sgl_used flag looks a little confuse to me, but i am not familiar with those detailed code.

Reply:
same as above.